### PR TITLE
JAMES-2646 list sources Cassandra table and DAO 

### PR DESCRIPTION
**Pull request by Arsnael**

### CODE REVIEW

File: server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java [hunk 1]
```diff
@@ -240,6 +240,11 @@ public Type getType() {
             return type;
         }
 
+        @Override
+        public String getMapping() {
```
Comment (chibenwa) [inline server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java]: I propose `getMappingValue`

File: server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraMappingsSourcesDAO.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    Mono<Void> removeMapping(Mapping mapping, MappingSource source) {
+        return executor.executeVoidReactor(deleteStatement.bind()
+            .setString(MAPPING_TYPE, mapping.getType().asPrefix())
+            .setString(MAPPING_VALUE, mapping.getMapping())
+            .setString(SOURCE, source.asMailAddressString()));
+    }
+
+    Mono<List<MappingSource>> retrieveSources(Mapping mapping) {
```
Comment (chibenwa) [inline server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraMappingsSourcesDAO.java]: What you are looking for is a flux.

File: server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraRRTModule.java [hunk 3]
```diff
@@ -37,5 +39,14 @@
             .addPartitionKey(CassandraRecipientRewriteTableTable.USER, text())
             .addClusteringColumn(CassandraRecipientRewriteTableTable.DOMAIN, text())
             .addClusteringColumn(CassandraRecipientRewriteTableTable.MAPPING, text()))
+        .table(CassandraMappingsSourcesTable.TABLE_NAME)
+        .comment("Holds mapping sources.")
```
Comment (chibenwa) [inline server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraRRTModule.java]: We have to e explaining it is a projection table for retrieving sources associated mappings.

File: server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraRecipientRewriteTable.java [hunk 4]
```diff
@@ -30,41 +31,55 @@
...
+    public CassandraRecipientRewriteTable(CassandraRecipientRewriteTableDAO cassandraRecipientRewriteTableDAO, CassandraMappingsSourcesDAO cassandraMappingsSourcesDAO) {
+        this.cassandraRecipientRewriteTableDAO = cassandraRecipientRewriteTableDAO;
+        this.cassandraMappingsSourcesDAO = cassandraMappingsSourcesDAO;
     }
 
     @Override
     public void addMapping(MappingSource source, Mapping mapping) {
-        dao.addMapping(source, mapping).block();
+        cassandraRecipientRewriteTableDAO.addMapping(source, mapping).block();
+        cassandraMappingsSourcesDAO.addMapping(mapping, source).block();
```
Comment (chibenwa) [inline server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraRecipientRewriteTable.java]: We should be chaining the operations, then calling a single time `.block()`

File: server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraRecipientRewriteTable.java [hunk 5]
```diff
@@ -30,41 +31,55 @@
...
         return OptionalUtils.orSuppliers(
-            () -> dao.retrieveMappings(MappingSource.fromUser(user, domain)).blockOptional(),
-            () -> dao.retrieveMappings(MappingSource.fromDomain(domain)).blockOptional())
+            () -> cassandraRecipientRewriteTableDAO.retrieveMappings(MappingSource.fromUser(user, domain)).blockOptional(),
+            () -> cassandraRecipientRewriteTableDAO.retrieveMappings(MappingSource.fromDomain(domain)).blockOptional())
                 .orElse(MappingsImpl.empty());
     }
+
+    @Override
+    public List<MappingSource> listSources(Mapping mapping) {
```
Comment (chibenwa) [inline server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraRecipientRewriteTable.java]: No not yet. We can read from this DAO only when we are sure that all the data are contained in this DAO

File: server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraMappingsSourcesDAO.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        return executor.executeReactor(retrieveSourcesStatement.bind()
+            .setString(MAPPING_TYPE, mapping.getType().asPrefix())
+            .setString(MAPPING_VALUE, mapping.getMapping()))
+            .map(resultSet -> cassandraUtils.convertToStream(resultSet)
+                .map(row -> row.getString(SOURCE))
+                .map(MappingSource::parse)
+                .collect(Guavate.toImmutableList()));
```
Comment (blackheaven) [inline server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraMappingsSourcesDAO.java]: `ImmutableList`

Comment (chibenwa) [reply server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraMappingsSourcesDAO.java]: No, as always, we have to keep using Guavate as the feature is experimental in Guava cf @mbaechler

Comment (Arsnael) [reply server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraMappingsSourcesDAO.java]: No List anymore anyway, switched to a `Flux` !

File: server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraRRTModule.java [hunk 7]
```diff
@@ -40,7 +40,7 @@
             .addClusteringColumn(CassandraRecipientRewriteTableTable.DOMAIN, text())
             .addClusteringColumn(CassandraRecipientRewriteTableTable.MAPPING, text()))
         .table(CassandraMappingsSourcesTable.TABLE_NAME)
-        .comment("Holds mapping sources.")
+        .comment("Projection table for retrieving sources associated mappings.")
```
Comment (chibenwa) [inline server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraRRTModule.java]: `associated with given mappings`

**Pull request closed without merging**